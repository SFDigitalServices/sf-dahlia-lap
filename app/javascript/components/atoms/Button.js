import React from 'react'

import classNames from 'classnames'
import PropTypes from 'prop-types'

import { toPx } from 'utils/cssUtils'

const buttonSizeStyles = (clearVerticalPadding, minWidth) => {
  const widthStyle = minWidth
    ? {
        minWidth: toPx(minWidth)
      }
    : {}

  const heightStyle = clearVerticalPadding
    ? {
        paddingTop: 0,
        paddingBottom: 0
      }
    : {}

  return {
    ...heightStyle,
    ...widthStyle
  }
}

const buttonChildrenStyles = () => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  width: '100%'
})

const textStyles = (hasLeftIcon, hasRightIcon) => ({
  marginTop: '1rem',
  marginBottom: '1rem',
  marginRight: hasRightIcon && '1rem',
  marginLeft: hasLeftIcon && '1rem',
  flexGrow: 1
})

const iconWrapperStyles = () => ({
  flexShrink: 0,

  // we need to override font-size because of a css button selector that sets
  // the font size for all children. This is a problem because it adds 3px
  // to any icon wrapper.
  fontSize: 0
})

const wrapWithStyle = (node, style) => node && <span style={style}>{node}</span>

const Button = ({
  classes = null,
  iconLeft = null,
  iconRight = null,
  noBottomMargin,
  paddingHorizontal = 'normal',
  small = false,
  tertiary = false,
  text = null,
  tiny = false,
  type = 'button',
  minWidthPx = null,
  style = {},
  ...rest
}) => {
  const btnClassNames = classNames(classes, {
    'tight-padding': paddingHorizontal === 'tight',
    'extra-padding': paddingHorizontal === 'extra',
    small: small,
    tertiary: tertiary,
    tiny: tiny,
    'margin-bottom-none': noBottomMargin
  })

  const hasText = !!text

  return (
    <button
      className={btnClassNames}
      type={type}
      style={{ ...buttonSizeStyles(hasText, minWidthPx), ...style }}
      {...rest}
    >
      <div style={buttonChildrenStyles()}>
        {wrapWithStyle(iconLeft, iconWrapperStyles())}
        {wrapWithStyle(text, textStyles(!!iconLeft, !!iconRight))}
        {wrapWithStyle(iconRight, iconWrapperStyles())}
      </div>
    </button>
  )
}

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'submit']),
  noBottomMargin: PropTypes.bool,
  classes: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  iconLeft: PropTypes.node,
  iconRight: PropTypes.node,
  minWidthPx: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  paddingHorizontal: PropTypes.oneOf(['normal', 'tight', 'extra']),
  small: PropTypes.bool,
  style: PropTypes.object,
  tertiary: PropTypes.bool,
  text: PropTypes.node,
  tiny: PropTypes.bool
}

export default Button
