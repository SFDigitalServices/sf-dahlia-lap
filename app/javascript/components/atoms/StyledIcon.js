import React from 'react'
import PropTypes from 'prop-types'

import { toRem } from '~/utils/cssUtils'

const DEFAULT_ICON_SIZE_REM = '1.5rem'

const getSizeStyle = (sizeRem) => ({
  width: sizeRem,
  height: sizeRem
})

const ICON_STYLE_REGULAR = getSizeStyle(DEFAULT_ICON_SIZE_REM)
const ICON_STYLE_MEDIUM = getSizeStyle('1rem')
const ICON_STYLE_SMALL = getSizeStyle('.75rem')
const ICON_STYLE_TINY = getSizeStyle('.5rem')

const getCustomSizeStyle = (customSize) => {
  let customSizeRem = toRem(customSize)

  if (customSizeRem === null) {
    console.error(`Invalid size value: ${customSize}`)
    customSizeRem = DEFAULT_ICON_SIZE_REM
  }

  return getSizeStyle(customSizeRem)
}

const getIconStyle = (size, customSize) => {
  switch (size) {
    case 'default':
      return ICON_STYLE_REGULAR
    case 'small':
      return ICON_STYLE_SMALL
    case 'medium':
      return ICON_STYLE_MEDIUM
    case 'tiny':
      return ICON_STYLE_TINY
  }
}

const StyledIcon = ({ size = 'default', customSizeRem = null, icon }) => (
  <svg style={customSizeRem ? getCustomSizeStyle(customSizeRem) : getIconStyle(size)}>
    <use xlinkHref={`#i-${icon}`} />
  </svg>
)

StyledIcon.propTypes = {
  icon: PropTypes.string.isRequired,
  size: PropTypes.oneOf(['default', 'small', 'medium', 'tiny']),
  customSizeRem: PropTypes.string
}

export default StyledIcon
