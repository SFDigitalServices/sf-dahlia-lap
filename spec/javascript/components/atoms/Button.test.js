import React from 'react'

import { shallow } from 'enzyme'

import Button from 'components/atoms/Button'
import StyledIcon from 'components/atoms/StyledIcon'

import { findWithText } from '../../testUtils/wrapperUtil'

describe('Button', () => {
  test('it renders a button with the correct default properties', () => {
    const wrapper = shallow(<Button text='This is a button' />)
    const selector = 'button'
    expect(wrapper.find(selector)).toHaveLength(1)
    expect(wrapper.find(selector).props().type).toEqual('button')
    expect(wrapper.find(selector).hasClass('margin-bottom-none')).toBeFalsy()
    expect(wrapper.find(StyledIcon)).toHaveLength(0)
  })

  test('it renders a button with the correct properties', () => {
    const selector = 'button'
    const wrapper = shallow(
      <Button
        type='submit'
        text='This is a button'
        small
        tiny
        tertiary
        noBottomMargin
        classes={['additional-class-one', 'additional-class-two']}
        data-other-prop='abc'
      />
    )
    const buttonWrapper = wrapper.find(selector)
    const buttonProps = buttonWrapper.props()
    expect(buttonWrapper).toHaveLength(1)
    expect(buttonProps.type).toEqual('submit')
    expect(buttonWrapper.text()).toEqual('This is a button')
    expect(buttonWrapper.hasClass('small')).toBeTruthy()
    expect(buttonWrapper.hasClass('tiny')).toBeTruthy()
    expect(buttonWrapper.hasClass('tertiary')).toBeTruthy()
    expect(buttonWrapper.hasClass('margin-bottom-none')).toBeTruthy()
    expect(buttonWrapper.hasClass('additional-class-one')).toBeTruthy()
    expect(buttonWrapper.hasClass('additional-class-two')).toBeTruthy()
    expect(buttonProps['data-other-prop']).toEqual('abc')
    expect(wrapper.find(StyledIcon)).toHaveLength(0)
  })

  describe('with an icon to the left', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(
        <Button text='This is a button' iconLeft={<StyledIcon icon='list-unordered' />} />
      )
    })

    test('should render a StyledIcon', () => {
      expect(wrapper.find(StyledIcon)).toHaveLength(1)
    })

    test('should render the StyledIcon before the text', () => {
      const firstChild = wrapper.childAt(0).childAt(0)
      const secondChild = wrapper.childAt(0).childAt(1)
      expect(firstChild.find(StyledIcon)).toHaveLength(1)
      expect(secondChild.find(StyledIcon)).toHaveLength(0)
      expect(firstChild.text()).not.toContain('This is a button')
      expect(secondChild.text()).toEqual('This is a button')
    })

    test('should add left margin to the text', () => {
      const textChild = wrapper.childAt(0).childAt(1)

      expect(textChild.props().style.marginLeft).toEqual('1rem')
      expect(textChild.props().style.marginRight).toBeFalsy()
    })
  })

  describe('with an icon to the right', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(
        <Button text='This is a button' iconRight={<StyledIcon icon='list-unordered' />} />
      )
    })

    test('should render a StyledIcon', () => {
      expect(wrapper.find(StyledIcon)).toHaveLength(1)
    })

    test('should render the StyledIcon after the text', () => {
      const firstChild = wrapper.childAt(0).childAt(0)
      const secondChild = wrapper.childAt(0).childAt(1)
      expect(firstChild.find(StyledIcon)).toHaveLength(0)
      expect(secondChild.find(StyledIcon)).toHaveLength(1)
      expect(firstChild.text()).toEqual('This is a button')
      expect(secondChild.text()).not.toContain('This is a button')
    })

    test('should add right margin to the text', () => {
      const textChild = wrapper.childAt(0).childAt(0)

      expect(textChild.props().style.marginRight).toEqual('1rem')
      expect(textChild.props().style.marginLeft).toBeFalsy()
    })
  })

  describe('with two icons', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(
        <Button
          text='This is a button'
          iconLeft={<StyledIcon icon='list-unordered' />}
          iconRight={<StyledIcon icon='list-unordered' />}
        />
      )
    })

    test('should render two StyledIcons', () => {
      expect(wrapper.find(StyledIcon)).toHaveLength(2)
    })

    test('should render the StyledIcon on either side of the text', () => {
      const firstChild = wrapper.childAt(0).childAt(0)
      const secondChild = wrapper.childAt(0).childAt(1)
      const thirdChild = wrapper.childAt(0).childAt(2)

      expect(firstChild.find(StyledIcon)).toHaveLength(1)
      expect(secondChild.find(StyledIcon)).toHaveLength(0)
      expect(thirdChild.find(StyledIcon)).toHaveLength(1)

      expect(firstChild.text()).not.toContain('This is a button')
      expect(secondChild.text()).toEqual('This is a button')
      expect(thirdChild.text()).not.toContain('This is a button')
    })

    test('should add right and left margin to the text', () => {
      const textChild = wrapper.childAt(0).childAt(1)

      expect(textChild.props().style.marginRight).toEqual('1rem')
      expect(textChild.props().style.marginLeft).toEqual('1rem')
    })
  })

  describe('when text is added', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<Button text='This is a button' />)
    })

    test('should clear vertical padding at the button level', () => {
      const buttonStyle = wrapper.find('button').props().style
      expect(buttonStyle.paddingTop).toEqual(0)
      expect(buttonStyle.paddingBottom).toEqual(0)
    })

    test('should add vertical padding to the text element', () => {
      const textChildWrapper = wrapper.childAt(0).childAt(0)
      const textStyle = textChildWrapper.props().style

      expect(textStyle.marginTop).toEqual('1rem')
      expect(textStyle.marginBottom).toEqual('1rem')
    })
  })

  describe('when text is empty', () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(
        <Button
          iconLeft={<StyledIcon icon='list-unordered' />}
          iconRight={<StyledIcon icon='list-unordered' />}
        />
      )
    })

    test('should not clear vertical padding at the button level', () => {
      const buttonStyle = wrapper.find('button').props().style
      expect(buttonStyle.paddingTop).toBeUndefined()
      expect(buttonStyle.paddingBottom).toBeUndefined()
    })
  })

  describe('textAlign', () => {
    const BUTTON_TEXT = 'test_button_text'

    test('renders left aligned text properly', () => {
      const wrapper = shallow(<Button text={BUTTON_TEXT} textAlign='left' />)

      expect(findWithText(wrapper, 'span', BUTTON_TEXT).props().style.textAlign).toEqual('left')
    })

    test('renders right aligned text properly', () => {
      const wrapper = shallow(<Button text={BUTTON_TEXT} textAlign='right' />)

      expect(findWithText(wrapper, 'span', BUTTON_TEXT).props().style.textAlign).toEqual('right')
    })

    test('renders center aligned text properly', () => {
      const wrapper = shallow(<Button text={BUTTON_TEXT} textAlign='center' />)

      expect(findWithText(wrapper, 'span', BUTTON_TEXT).props().style.textAlign).toEqual('center')
    })

    test('renders center aligned text by default', () => {
      const wrapper = shallow(<Button text={BUTTON_TEXT} />)

      expect(findWithText(wrapper, 'span', BUTTON_TEXT).props().style.textAlign).toEqual('center')
    })
  })
})
