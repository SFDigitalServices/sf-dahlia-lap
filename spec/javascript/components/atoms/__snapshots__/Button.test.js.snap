// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button textAlign renders center aligned text by default 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; overflow: auto; flex-grow: 1; text-align: center;"
      >
        test_button_text
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button textAlign renders center aligned text properly 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; overflow: auto; flex-grow: 1; text-align: center;"
      >
        test_button_text
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button textAlign renders left aligned text properly 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; overflow: auto; flex-grow: 1; text-align: left;"
      >
        test_button_text
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button textAlign renders right aligned text properly 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; overflow: auto; flex-grow: 1; text-align: right;"
      >
        test_button_text
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button when text is added should clear vertical padding at the text and button 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; overflow: auto; flex-grow: 1; text-align: center;"
      >
        This is a button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button when text is empty should not clear vertical padding at the button level 1`] = `
<DocumentFragment>
  <button
    class=""
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="flex-shrink: 0; font-size: 0px;"
      >
        <svg
          style="width: 1.5rem; height: 1.5rem;"
        >
          <use
            xlink:href="#i-list-unordered"
          />
        </svg>
      </span>
      <span
        style="flex-shrink: 0; font-size: 0px;"
      >
        <svg
          style="width: 1.5rem; height: 1.5rem;"
        >
          <use
            xlink:href="#i-list-unordered"
          />
        </svg>
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button with an icon to the left should render the StyledIcon before the text 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="flex-shrink: 0; font-size: 0px;"
      >
        <svg
          data-testid="list-unordered"
          style="width: 1.5rem; height: 1.5rem;"
        >
          <use
            xlink:href="#i-list-unordered"
          />
        </svg>
      </span>
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; margin-left: 1rem; overflow: auto; flex-grow: 1; text-align: center;"
      >
        This is a button
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button with an icon to the right should render the StyledIcon after the text 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="margin-top: 1rem; margin-bottom: 1rem; margin-right: 1rem; overflow: auto; flex-grow: 1; text-align: center;"
      >
        This is a button
      </span>
      <span
        style="flex-shrink: 0; font-size: 0px;"
      >
        <svg
          data-testid="list-unordered"
          style="width: 1.5rem; height: 1.5rem;"
        >
          <use
            xlink:href="#i-list-unordered"
          />
        </svg>
      </span>
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button with two icons should render the StyledIcon on either side of the text 1`] = `
<DocumentFragment>
  <button
    class=""
    style="padding-top: 0px; padding-bottom: 0px;"
    type="button"
  >
    <div
      style="display: flex; align-items: center; justify-content: space-between; width: 100%;"
    >
      <span
        style="flex-shrink: 0; font-size: 0px;"
      >
        <svg
          data-testid="list-unordered"
          style="width: 1.5rem; height: 1.5rem;"
        >
          <use
            xlink:href="#i-list-unordered"
          />
        </svg>
      </span>
      <span
        style="margin: 1rem 1rem 1rem 1rem; overflow: auto; flex-grow: 1; text-align: center;"
      >
        This is a button
      </span>
      <span
        style="flex-shrink: 0; font-size: 0px;"
      >
        <svg
          data-testid="list-unordered"
          style="width: 1.5rem; height: 1.5rem;"
        >
          <use
            xlink:href="#i-list-unordered"
          />
        </svg>
      </span>
    </div>
  </button>
</DocumentFragment>
`;
